cpp/src/arrow/io/caching.h

```diff
@@ -99,6 +100,9 @@ class ARROW_EXPORT ReadRangeCache {
   /// \brief Read a range previously given to Cache().
   Result<std::shared_ptr<Buffer>> Read(ReadRange range);
 
+  /// \brief Wait until all ranges added so far have been cached.
+  Future<> Wait();
```

Comment:
Hmm... Wouldn't it be better to expose a `ReadAsync` equivalent to `Read`?
```c++
Future<std::shared_ptr<Buffer>> Read(ReadRange range);
```

Reply:
The idea here is to cleanly separate out the I/O operations so you can schedule them on different executors. With a ReadAsync, you wouldn't be able to accomplish that unless you went and refactored the entire Parquet reader so that it was fully async-aware.

Reply:
Fair enough :-)